File: ios/Sources/AICLICompanion/ClaudeOutputParser.swift
import Foundation

/// Parser for Claude's structured output
/// Detects and extracts user-facing content from Claude's JSON responses
@available(iOS 16.0, macOS 13.0, *)
class ClaudeOutputParser {
    
    /// Attempts to parse Claude's JSON output and extract the user-facing content
    /// - Parameter text: Raw text that might contain JSON
    /// - Returns: Parsed result with extracted content or nil
    static func parseClaudeOutput(_ text: String) -> (content: String, metadata: [String: Any])? {
        guard !text.isEmpty else { return nil }
        
        // Try to find JSON blocks in the text
        let jsonPattern = #"\{[\s\S]*\}"#
        guard let regex = try? NSRegularExpression(pattern: jsonPattern, options: []),
              let match = regex.firstMatch(in: text, options: [], range: NSRange(location: 0, length: text.utf16.count)),
              let jsonRange = Range(match.range, in: text) else {
            return nil
        }
        
        let jsonString = String(text[jsonRange])
        
        // Try to parse the JSON
        guard let jsonData = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] else {
            return nil
        }
        
        // Check if this looks like Claude's structured output
        guard isClaudeStructuredOutput(json) else { return nil }
        
        // Extract the user-facing content
        if let content = extractUserContent(from: json) {
            let metadata: [String: Any] = [
                "type": json["type"] as? String ?? "unknown",
                "hasThinking": json["thinking"] != nil,
                "hasToolUse": (json["tool_calls"] != nil || json["tools_used"] != nil)
            ]
            return (content: content, metadata: metadata)
        }
        
        return nil
    }
    
    /// Check if the parsed object looks like Claude's structured output
    private static func isClaudeStructuredOutput(_ json: [String: Any]) -> Bool {
        // Common patterns in Claude's JSON output
        return json["content"] != nil ||
               json["result"] != nil ||
               json["answer"] != nil ||
               json["thinking"] != nil ||
               json["reasoning"] != nil ||
               json["tool_calls"] != nil ||
               json["tools_used"] != nil ||
               json["type"] != nil
    }
    
    /// Extract user-facing content from Claude's structured output
    private static func extractUserContent(from json: [String: Any]) -> String? {
        // Priority order for content extraction
        if let content = json["content"] as? String {
            return content
        }
        
        if let result = json["result"] as? String {
            return result
        }
        
        if let answer = json["answer"] as? String {
            return answer
        }
        
        if let message = json["message"] as? String {
            return message
        }
        
        if let text = json["text"] as? String {
            return text
        }
        
        if let response = json["response"] as? String {
            return response
        }
        
        // For tool use, create a summary
        if let toolCalls = json["tool_calls"] as? [[String: Any]] ?? json["tools_used"] as? [[String: Any]] {
            let toolNames = toolCalls.compactMap { $0["name"] as? String ?? $0["tool"] as? String }
            if !toolNames.isEmpty {
                return "Using tools: \(toolNames.joined(separator: ", "))"
            }
        }
        
        return nil
    }
}

File: macos-app/AICLICompanionHost/Resources/server/src/services/claude-output-parser.js
/**
 * Parser for Claude's structured output
 * Detects and extracts user-facing content from Claude's JSON responses
 */

/**
 * Attempts to parse Claude's JSON output and extract the user-facing content
 * @param {string} text - Raw text that might contain JSON
 * @returns {Object|null} Parsed result with { isJson: boolean, content: string, metadata: Object }
 */
export function parseClaudeOutput(text) {
  if (!text || typeof text !== 'string') {
    return null;
  }

  // Try to detect JSON blocks in the text
  const jsonMatches = text.match(/\{[\s\S]*\}/g);
  if (!jsonMatches) {
    return null;
  }

  for (const jsonMatch of jsonMatches) {
    try {
      const parsed = JSON.parse(jsonMatch);

      // Check if this looks like Claude's structured output
      if (isClaudeStructuredOutput(parsed)) {
        // Extract the user-facing content
        const content = extractUserContent(parsed);
        return {
          isJson: true,
          content,
          metadata: {
            type: parsed.type || 'unknown',
            hasThinking: !!parsed.thinking,
            hasToolUse: !!(parsed.tool_calls || parsed.tools_used),
          },
        };
      }
    } catch (e) {
      // Not valid JSON, continue
    }
  }

  return null;
}

/**
 * Check if the parsed object looks like Claude's structured output
 */
function isClaudeStructuredOutput(obj) {
  // Common patterns in Claude's JSON output
  return (
    obj &&
    typeof obj === 'object' &&
    // Has content field (main response)
    (obj.content ||
      // Has result field
      obj.result ||
      // Has answer field
      obj.answer ||
      // Has thinking/reasoning fields
      obj.thinking ||
      obj.reasoning ||
      // Has tool-related fields
      obj.tool_calls ||
      obj.tools_used ||
      // Has type field indicating structured response
      obj.type)
  );
}

/**
 * Extract user-facing content from Claude's structured output
 */
function extractUserContent(parsed) {
  // Priority order for content extraction
  if (parsed.content) {
    return parsed.content;
  }

  if (parsed.result) {
    return parsed.result;
  }

  if (parsed.answer) {
    return parsed.answer;
  }

  // If there's a message field
  if (parsed.message) {
    return parsed.message;
  }

  // If there's text field
  if (parsed.text) {
    return parsed.text;
  }

  // If there's a response field
  if (parsed.response) {
    return parsed.response;
  }

  // For tool use, create a summary
  if (parsed.tool_calls || parsed.tools_used) {
    const tools = parsed.tool_calls || parsed.tools_used;
    if (Array.isArray(tools) && tools.length > 0) {
      return `Using tools: ${tools.map((t) => t.name || t.tool).join(', ')}`;
    }
  }

  // Fallback: try to find any string content
  for (const key of Object.keys(parsed)) {
    if (
      typeof parsed[key] === 'string' &&
      key !== 'thinking' &&
      key !== 'reasoning' &&
      key !== 'type' &&
      parsed[key].length > 0
    ) {
      return parsed[key];
    }
  }

  return null;
}

/**
 * Clean up text by removing JSON blocks that have been parsed
 */
export function removeJsonFromText(text, jsonContent) {
  if (!text || !jsonContent) {
    return text;
  }

  // Escape special regex characters in the JSON content
  const escaped = jsonContent.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  return text.replace(new RegExp(escaped, 'g'), '').trim();
}
